# Pima Data cleaning



In [1]:
# Import Modules Here
import pandas as pd

## Part 1: Read the Data into Jupyter

In [2]:
# Read the data into a dataframe named pima_df.
# YOUR CODE HERE
pima_df = pd.read_csv('./data/diabetes_uncleaned_data.csv')
display(pima_df)

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
0             3.0    167.0            39             28      545  53.0   
1            13.0    151.0           107             38    10000   9.2   
2             8.0     59.0           109             18      739  59.1   
3            11.0    132.0            83             11      376   4.1   
4            10.0    165.0            48             37      737  39.4   
...           ...      ...           ...            ...      ...   ...   
999           6.0    110.0            26             27      590  20.8   
1000          6.0      NaN            10              4      640  55.7   
1001          4.0     80.0            24              6      268  48.9   
1002          3.0    142.0         Error             10      299   4.8   
1003          6.0    151.0            60             26      362  43.4   

      DiabetesPedigreeFunction    Age Outcome  
0                        2.447  125.0   FALSE  
1                        0.507  148.0   FALSE  
2                        1.769   91.0    TRUE  
3                        2.772    NaN    TRUE  
4                        0.729   73.0   FALSE  
...                        ...    ...     ...  
999                      1.185   95.0   FALSE  
1000                     2.483   79.0   FALSE  
1001                     0.989   81.0    TRUE  
1002                     2.411   71.0    TRUE  
1003                     2.027   23.0   FALSE  

[1004 rows x 9 columns]

## Part 2: Missing Data

In [3]:
#Question 1:  Find Columns with missing data
# YOUR CODE HERE
print(pima_df.isna().sum())
print(pima_df.isnull().sum())

Pregnancies                 114
Glucose                      16
BloodPressure                22
SkinThickness                 0
Insulin                       0
BMI                           1
DiabetesPedigreeFunction      0
Age                          22
Outcome                       0
dtype: int64
Pregnancies                 114
Glucose                      16
BloodPressure                22
SkinThickness                 0
Insulin                       0
BMI                           1
DiabetesPedigreeFunction      0
Age                          22
Outcome                       0
dtype: int64


In [4]:
# Question 2:  Write code to remove rows with missing data and save the new dataframe as pima_missing_fixed_df
# YOUR CODE HERE
pima_missing_fixed_df = pima_df.dropna(how='any')
display(pima_missing_fixed_df)

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
0             3.0    167.0            39             28      545  53.0   
1            13.0    151.0           107             38    10000   9.2   
2             8.0     59.0           109             18      739  59.1   
4            10.0    165.0            48             37      737  39.4   
5          -100.0    142.0            88             39       82  42.7   
...           ...      ...           ...            ...      ...   ...   
998           6.0    110.0            26             27      590  20.8   
999           6.0    110.0            26             27      590  20.8   
1001          4.0     80.0            24              6      268  48.9   
1002          3.0    142.0         Error             10      299   4.8   
1003          6.0    151.0            60             26      362  43.4   

      DiabetesPedigreeFunction    Age Outcome  
0                        2.447  125.0   FALSE  
1                        0.507  148.0   FALSE  
2                        1.769   91.0    TRUE  
4                        0.729   73.0   FALSE  
5                        1.369  148.0    TRUE  
...                        ...    ...     ...  
998                      1.185   95.0   FALSE  
999                      1.185   95.0   FALSE  
1001                     0.989   81.0    TRUE  
1002                     2.411   71.0    TRUE  
1003                     2.027   23.0   FALSE  

[840 rows x 9 columns]

In [5]:
# Question 3:  Print out the number of rows in the dataframe.
# YOUR CODE HERE
print(pima_df['Insulin'].count())
print(pima_missing_fixed_df['Insulin'].count())

1004
840


### Question 4:  Discuss why this could be a problem.  What other methods could you use in this situation? 

This way of removing rows with missing values removes a large chunk, roughly 25% of our data from the data set. In this case, it may be better practice to remove the rows who have missing data only in the key columns we need for analysis. For example, we could first remove rows who have no value for Glucose. Given this is a diabetes study, we need to have accurate glucose values to perform any analysis related to diabetes. If we do it this way, we only remove 16 records from the data set versus ~200. We could continue going in order of column importance, removing data this way in chunks with more precision.

## Part 3: Duplicated Data

In [6]:
# Question 1: find the rows of duplicated data in the 
# pima_missing_fixed_df and print them.  
pima_missing_fixed_df[pima_missing_fixed_df.duplicated()]


Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
925           6.0     39.0            46              2      287  40.4   
926          10.0     67.0            39             12      168  53.4   
999           6.0    110.0            26             27      590  20.8   
1003          6.0    151.0            60             26      362  43.4   

      DiabetesPedigreeFunction   Age Outcome  
925                      2.394  60.0    TRUE  
926                      0.162  71.0   FALSE  
999                      1.185  95.0   FALSE  
1003                     2.027  23.0   FALSE

In [7]:
# Question 2: Write code to remove the duplicate rows and save the new dataframe as pima_dedupped_df
print(pima_missing_fixed_df['Insulin'].count())
pima_dedupped_df = pima_missing_fixed_df.drop_duplicates(keep='first')
print(pima_dedupped_df['Insulin'].count())

840
836


In [8]:
# Question 3:  Print out the number of rows in the dataframe.
print(pima_dedupped_df['Insulin'].count())

836


## Part 4: Mis-typed Columns/Broken Data


In [9]:
# Question 1: find the columns where most of the data seems to be one type of data, and there seems to be a data error. In the markdown, after exploring in code, add a markdown cell discussing which columns have the data error. 
display(pima_dedupped_df)
pima_dedupped_df.info()
print(pima_dedupped_df['BloodPressure'].value_counts())
print(pima_dedupped_df['BMI'].value_counts())
display(pima_dedupped_df[pima_dedupped_df['BMI'] == 'Error'])
print(pima_dedupped_df['Outcome'].value_counts())

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
0             3.0    167.0            39             28      545  53.0   
1            13.0    151.0           107             38    10000   9.2   
2             8.0     59.0           109             18      739  59.1   
4            10.0    165.0            48             37      737  39.4   
5          -100.0    142.0            88             39       82  42.7   
...           ...      ...           ...            ...      ...   ...   
996           6.0    117.0            91             28      212   1.0   
997          13.0    163.0            10             37      457   8.0   
998           6.0    110.0            26             27      590  20.8   
1001          4.0     80.0            24              6      268  48.9   
1002          3.0    142.0         Error             10      299   4.8   

      DiabetesPedigreeFunction    Age Outcome  
0                        2.447  125.0   FALSE  
1                        0.507  148.0   FALSE  
2                        1.769   91.0    TRUE  
4                        0.729   73.0   FALSE  
5                        1.369  148.0    TRUE  
...                        ...    ...     ...  
996                      1.969  137.0   FALSE  
997                      2.400  125.0    TRUE  
998                      1.185   95.0   FALSE  
1001                     0.989   81.0    TRUE  
1002                     2.411   71.0    TRUE  

[836 rows x 9 columns]

<class 'pandas.core.frame.DataFrame'>
Index: 836 entries, 0 to 1002
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               836 non-null    float64
 1   Glucose                   836 non-null    float64
 2   BloodPressure             836 non-null    object 
 3   SkinThickness             836 non-null    int64  
 4   Insulin                   836 non-null    int64  
 5   BMI                       836 non-null    object 
 6   DiabetesPedigreeFunction  836 non-null    float64
 7   Age                       836 non-null    float64
 8   Outcome                   836 non-null    object 
dtypes: float64(4), int64(2), object(3)
memory usage: 65.3+ KB
BloodPressure
26       15
82       15
27       15
96       15
81       13
         ..
107       4
13        3
10000     2
Error     2
19        1
Name: count, Length: 113, dtype: int64
BMI
7.0     6
17.4    6
59.6    6
15.1    6
34.

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin    BMI  \
643          5.0    189.0             8             32        2  Error   

     DiabetesPedigreeFunction    Age Outcome  
643                     2.916  107.0    TRUE

Outcome
FALSE    441
TRUE     394
ERROR      1
Name: count, dtype: int64


In [10]:
# Question 2: Remove the rows with data errors - should remove 4 records from pima_dedupped_df
pima_remove_error_1_df = pima_dedupped_df[(pima_dedupped_df['BloodPressure'] != 'Error') & (pima_dedupped_df['Outcome'] != 'ERROR') & (pima_dedupped_df['BMI'] != 'Error')]
print(pima_remove_error_1_df['Insulin'].count())

832


In [11]:
# Question 3: After the data is fixed in your columns change the columns to the correct type and save this as the pima_fixed_columns_df.
pima_fixed_columns_df = pima_remove_error_1_df.copy()
pima_fixed_columns_df['BloodPressure'] = pima_fixed_columns_df['BloodPressure'].astype('int')
pima_fixed_columns_df['BMI'] = pima_fixed_columns_df['BMI'].astype('float')

def convert_bool(x):
    print(x)
    if x == "FALSE":
        return False
    elif x == "TRUE":
        return True

pima_fixed_columns_df['Outcome'] = pima_fixed_columns_df['Outcome'].apply(convert_bool)

pima_fixed_columns_df['Outcome'] = pima_fixed_columns_df['Outcome'].astype('bool')


FALSE
FALSE
TRUE
FALSE
TRUE
FALSE
FALSE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
FALSE
TRUE
FALSE
FALSE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
TRUE
FALSE
FALSE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
TRUE
TRUE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
FALSE
TRUE
TRUE
FALSE
FALSE
FALSE
FALSE
FALSE
FALSE
TRUE
FALSE
TRUE
TRUE
TRUE
TRUE
TRUE
FALSE
FALSE
FALSE
TRUE
TRUE
FALSE
FALSE
TRUE
FALSE
TRUE
TRUE
FALSE
TRUE
TRUE
TRUE
FALSE
FALSE
FALSE
TRUE
FALSE
TRUE
FALSE
TRUE
TRUE
FALSE
TRUE
FALSE
FALSE
TRUE
FALSE
FALSE
FALSE
FALSE
TRUE
FALSE
TRUE
FALSE
FALSE
FALSE
TRUE
TRUE
TRUE
FALSE
TRUE
FALSE
TRUE
TRUE
TRUE
TRUE
TRUE
FALSE
TRUE
FALSE
FALSE
FALSE
FALSE
FALSE
FALSE
TRUE
TRUE
TRUE
FALSE
FALSE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
TRUE
TRUE
TRUE
TRUE
TRUE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
FALSE
FALSE
TRUE
FALSE
TRUE
TRUE
FALSE
TRUE
TRUE
FALSE
FALSE
TRUE
TRUE
FALSE
TRUE
TRUE
TRUE
TRUE
TRUE
FALSE
FALSE
FALSE
FALSE
FALSE
TRUE
TRUE
TRUE
FALSE
TRUE
FALSE
TRUE
TRUE
TRUE
FALSE
FALSE
TRUE
TRUE
TRUE
TRUE

In [12]:
# Question 4: Run pima_fixed_columns_df.info() to confirm the columns have changed.
pima_fixed_columns_df.info()
display(pima_fixed_columns_df)

<class 'pandas.core.frame.DataFrame'>
Index: 832 entries, 0 to 1001
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               832 non-null    float64
 1   Glucose                   832 non-null    float64
 2   BloodPressure             832 non-null    int32  
 3   SkinThickness             832 non-null    int64  
 4   Insulin                   832 non-null    int64  
 5   BMI                       832 non-null    float64
 6   DiabetesPedigreeFunction  832 non-null    float64
 7   Age                       832 non-null    float64
 8   Outcome                   832 non-null    bool   
dtypes: bool(1), float64(5), int32(1), int64(2)
memory usage: 56.1 KB


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             3.0    167.0             39             28      545  53.0   
1            13.0    151.0            107             38    10000   9.2   
2             8.0     59.0            109             18      739  59.1   
4            10.0    165.0             48             37      737  39.4   
5          -100.0    142.0             88             39       82  42.7   
...           ...      ...            ...            ...      ...   ...   
995           5.0     62.0          10000             13      412  48.7   
996           6.0    117.0             91             28      212   1.0   
997          13.0    163.0             10             37      457   8.0   
998           6.0    110.0             26             27      590  20.8   
1001          4.0     80.0             24              6      268  48.9   

      DiabetesPedigreeFunction    Age  Outcome  
0                        2.447  125.0    False  
1                        0.507  148.0    False  
2                        1.769   91.0     True  
4                        0.729   73.0    False  
5                        1.369  148.0     True  
...                        ...    ...      ...  
995                      0.267  101.0    False  
996                      1.969  137.0    False  
997                      2.400  125.0     True  
998                      1.185   95.0    False  
1001                     0.989   81.0     True  

[832 rows x 9 columns]

## Part 5: Outlier Detection and Removal


In [13]:
# Question 1: Print out the Outliers in each column in the pima_fixed_columns_df dataframe, use the IQR method of outlier detection.
def is_iqr_outlier(column):
    q1 = column.quantile(0.25)
    q3 = column.quantile(0.75)
    IQR = q3 - q1
    lower_bound = q1 - (1.5 * IQR)
    upper_bound = q3 + (1.5 * IQR)
    return ~column.between(lower_bound, upper_bound)

print("Pregnancies Outliers:")
display(pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["Pregnancies"])])
print("Glucose Outliers:")
display(pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["Glucose"])])
print("Blood Pressure Outliers:")
display(pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["BloodPressure"])])
print("Skin Thickness Outliers:")
display(pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["SkinThickness"])])
print("Insulin Outliers:")
display(pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["Insulin"])])
print("BMI Outliers:")
display(pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["BMI"])])
print("DiabetesPedigreeFunction Outliers:")
display(pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["DiabetesPedigreeFunction"])])
print("Age Outliers:")
display(pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["Age"])])




Pregnancies Outliers:


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
5       -100.0    142.0             88             39       82  42.7   

   DiabetesPedigreeFunction    Age  Outcome  
5                     1.369  148.0     True

Glucose Outliers:


Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

Blood Pressure Outliers:


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
649          3.0    155.0          10000             35      748  32.6   
995          5.0     62.0          10000             13      412  48.7   

     DiabetesPedigreeFunction    Age  Outcome  
649                     0.098   85.0     True  
995                     0.267  101.0    False

Skin Thickness Outliers:


Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

Insulin Outliers:


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  BMI  \
1         13.0    151.0            107             38    10000  9.2   

   DiabetesPedigreeFunction    Age  Outcome  
1                     0.507  148.0    False

BMI Outliers:


Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

DiabetesPedigreeFunction Outliers:


Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

Age Outliers:


Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

In [14]:
# Question 2: Use loc to remove outliers in each of the columns that have outliers, save this as pima_outlier_removed_df.
pima_outlier_removed_df = pima_fixed_columns_df[is_iqr_outlier(pima_fixed_columns_df["Pregnancies"]) == False]
print(pima_outlier_removed_df['Insulin'].count())
pima_outlier_removed_df = pima_outlier_removed_df[is_iqr_outlier(pima_fixed_columns_df["BloodPressure"]) == False]
print(pima_outlier_removed_df['Insulin'].count())
pima_outlier_removed_df = pima_outlier_removed_df[is_iqr_outlier(pima_fixed_columns_df["Insulin"]) == False]
pima_outlier_removed_df['Insulin'].count()
#Should be 4 records less than pima_fixed_columns_df because 4 outliers were found

831
829


C:\Users\brend\AppData\Local\Temp\ipykernel_20708\208433302.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pima_outlier_removed_df = pima_outlier_removed_df[is_iqr_outlier(pima_fixed_columns_df["BloodPressure"]) == False]
C:\Users\brend\AppData\Local\Temp\ipykernel_20708\208433302.py:6: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pima_outlier_removed_df = pima_outlier_removed_df[is_iqr_outlier(pima_fixed_columns_df["Insulin"]) == False]


828

In [15]:
# Question 3:  Print out the row count in the pima_outlier_removed_df and confirm this number is correct.
pima_outlier_removed_df['Insulin'].count()

828

Based on the data cleaning steps taken, 828 records is correct.